### PR TITLE
Remove TypeFieldMapper

### PR DESCRIPTION
**Pull request by romseygeek**
Now that nested documents have their own nested path metadata field in 8x, the 
`_type` field itself is entirely unnecessary.  This commit removes the `TypeFieldMapper`,
and adds some special-case logic to `QueryShardContext` to return a singleton instance
of `TypeFieldType` when somebody asks for the `_type` field to build queries.

### CODE REVIEW

**Comment by romseygeek** [comment]
@pgomulka @jakelandis I think we'll eventually want to hide the special case logic in `QueryShardContext` here in a v7 compatibility wrapper so that queries against the `_type` field are only parsed for v7 clients.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -74,6 +76,11 @@
 
 public class MapperService extends AbstractIndexComponent implements Closeable {
 
+    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(
+        LogManager.getLogger(QueryShardContext.class));
+    public static final String TYPES_DEPRECATION_MESSAGE = "[types removal] Using the _type field " +
+        "in queries and aggregations is deprecated, prefer to use a field instead.";
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: nit: If this is only applicable for v7 compatibility mode, should we cal this constant: `TYPES_DEPRECATION_MESSAGE_V7`?

Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: Should we remove `_type` field from [SORTED_META_FIELDS](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/mapper/MapperService.java#L113) constant?

File: server/src/main/java/org/elasticsearch/index/mapper/TypeFieldType.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+package org.elasticsearch.index.mapper;
+
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.plain.ConstantIndexFieldData;
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.search.aggregations.support.CoreValuesSourceType;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
+
+@Deprecated
+public final class TypeFieldType extends ConstantFieldType {
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/mapper/TypeFieldType.java]: Is this class only be used for V7 compatibility mode? may be also worth to put a comment specifying this.

**Comment by pgomulka** [comment]
if this PR can wait for https://github.com/elastic/elasticsearch/pull/53228 to be merged we can use the infrastructure to wrap v7 code and see that type related yml tests from 7.x are still passing

**Comment by romseygeek** [comment]
@pgomulka that's a good idea, we might be able to move things into the compatibility module

**Comment by romseygeek** [comment]
This has got very old, I'm going to close and open a new PR.

**Comment by joegallo** [comment]
Related to https://github.com/elastic/elasticsearch/pull/62838

**Pull request closed without merging**